---
title: "README"
author: "dnanto"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    keep_md: yes
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	warning = FALSE,
	message = FALSE,
	comment = ""
)
if (!exists("PATH")) PATH = system('source ~/.bash_profile && echo "$PATH"', intern = T)
Sys.setenv(PATH = PATH)
```

# Abstract

**ffdb** is a collection of scripts to work with flat-file sequence databases and help me complete my thesis.

Note: this repository is experimental...
Note: this is an Rmarkdown document and uses a mix of R and bash code chunks.

# Dependencies

## Environment

### PYTHONPATH

Append the directory that has the package folder to PYTHONPATH.

```{r}
if (!exists("PYTHONPATH")) PYTHONPATH = getwd()
Sys.setenv(PYTHONPATH = PYTHONPATH)
```

```{bash}
echo ${PYTHONPATH/~/\~}
```

### FFIDX

This is an optional environment variable similar to BLASTDB. It is a colon-separated list of absolute or relative paths to search for indexed sets of flat-files. The examples take advantage of this feature. Otherwise, it is possible to specify the path to the index to the script directly.

```{r}
Sys.setenv(FFIDX = file.path(getwd(), "data"))
```

```{bash}
echo ${FFIDX/~/\~}
```

## Note

* The Pipfile lists the Python requirements.
* The edirect programs should be in PATH; download them [here](ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/).
* For convenience, add the script root directory to PATH.
* Also, it is convenient to know the NCBI [search field descriptions](https://www.ncbi.nlm.nih.gov/books/NBK49540/).

# Scripts

## ffdb.py

The ```ffdb.py``` script searches NCBI and downloads the entire result set. The destination folder stores the [BGZip](http://www.htslib.org/doc/bgzip.html)-compressed flat-files. The [index_db](https://biopython.org/DIST/docs/api/Bio.SeqIO-module.html#index_db) method generates the [SQLite](https://www.sqlite.org/index.html) database in the same directory. The script also modifies the meta_data table to preserve NCBI search parameters.

```{bash}
ffdb.py -h
```

### Ex-1

Initialize a repository of indexed GenBank files. Search NCBI for all *Mimivirus* genomic DNA sequences.

```{bash}
ffdb.py data/315393 \
  -term 'txid315393[PORGN] AND biomol_genomic[PROP] NOT gbdiv_pat[PROP] NOT gbdiv_syn[PROP]' \
  -fmt gbk -post 100
```

Update the FFIDX environment variable.

List the results in the directory.

```{bash}
du -sh data/315393.*
```

### Ex-2

Update the same repository.

```{bash}
ffdb.py 315393
```

### Ex-2

Update the same repository, but disable the modified date limiter.

```{bash}
ffdb.py 315393 -no-mdat
```

## ffidx.py

The ```ffidx.py``` script creates and/or queries an indexed set of sequence files created via the [index_db](https://biopython.org/DIST/docs/api/Bio.SeqIO-module.html#index_db) method.

```{bash}
ffidx.py -h
```

### Ex-1

Create an index using multiple multi-GenBank files and query some accessions.

```{bash}
ls data/oantigen.[1-2].gbk.gz | \
  xargs ffidx.py data/oantigen.idx -acc AF390573.1 GU576499.1 -fmt-idx gb -filenames | \
  grep -A 2 \>
```

List the index.

```{bash}
du -sh data/oantigen.idx
```

### Ex-2

Same thing, except create the index temporarily in memory.

```{bash}
ls data/oantigen.[1-2].gbk.gz | \
  xargs ffidx.py ":memory:" -acc AF390573.1 GU576499.1 -fmt-idx gb -filenames | \
  grep -A 2 \>
```

### Ex-3

Dump all sequences as GenBank records.

```{bash}
ffidx.py oantigen -dump -fmt-o gb | grep ^DEFINITION
```

## ffcds.py

The ```ffcds.py``` script extract all CDS records from a GenBank file.

```{bash}
ffcds.py -h
```

### Ex-1

Extract all CDS records.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | ffcds.py - | grep -A 2 \> | head
```

## fffilter.py

The ```fffilter.py``` script filters for records with matching regex patterns and/or length bounds.

```{bash}
fffilter.py -h
```

### Query Syntax

The length syntax is the LEN keyword + operator + number or range. Choose an operator: <, <=, =, ==, >=, >, IN. The range format is two numbers separated by a dash.

The query syntax is the LIKE keyword + property + regex. The property is a [SeqRecord](http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html) property.

To negate the results, prepend an N to the keyword: NLEN, NLIKE.

TODO: allow access to subproperties and fields, such as annotations and qualifiers...

### Ex-1

Filter sequences based on a regex pattern.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fffilter.py - "LIKE description CO[0-9]+ " -fmt-in gb | grep ^DEFINITION
```

Negate the previous search.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fffilter.py - "NLIKE description CO[0-9]+ " -fmt-in gb | grep ^DEFINITION
```

### Ex-2

Filter sequences based on length bounds: 45000 bp +/- 5000.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fffilter.py - "LEN IN 40000-50000" -fmt-in gb | grep ^LOCUS
```

Negate the previous search.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fffilter.py - "NLEN IN 40000-50000" -fmt-in gb | grep ^LOCUS
```

## fflen.py

The ```fflen.py``` script computes the length for each sequence in the flat-file.

```{bash}
fflen.py -h
```

### Ex-1

List the length of each record in the file.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fflen.py - -fmt gb
```

### Ex-2

Compute the length summary statistics.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | fflen.py - -fmt gb -stat
```

## ffgbk2taxmap.py

The ```ffgbk2taxmap.py``` script generates a taxonomy mapping from a GenBank file suitable for use with the ```makeblastdb``` command.

```{bash}
ffgbk2taxmap.py -h
```

### Ex-1

Generate a taxonomy mapping from a GenBank file.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | ffgbk2taxmap.py - 
```

## ffuseq.py

The ```ffuseq.py``` script computes the unique set of sequences and an optional tab-separated mapping file.

```{bash}
ffuseq.py -h
```

### Ex-1

This example creates a file stream with duplicate records. Output the unique set and a mapping file.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | ffuseq.py - -fmt gb -map data/urec.tsv > data/urec.gbk
```

```grep``` Unique record LOCUS tags.

```{bash}
grep ^LOCUS data/urec.gbk
```

```cat``` the unique record mapping file.

```{bash}
cat data/urec.tsv
```

## ffgbksrc.py

The ```ffgbksrc``` command queries GenBank metadata from the **source** feature.

```{bash}
ffgbksrc.py -h
```

### Key Syntax

The key syntax is key:default:join, where key is the key qualifier, default is the default value if the key is missing, and join is a character to join multiple values under the same key. The default and join specifiers are optional.

### Ex-1

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz | ffgbksrc.py - strain db_xref:0:,
```

## fffnagff2gbk.py

The ```fffnagff2gbk.py``` script combines FASTA and GFF3 files to create GenBank files with feature annotations.

```{bash}
fffnagff2gbk.py -h
```

### Ex-1

...

## ffkmer.py

The ```ffkmer``` script k-merizes sequences into separate records.

```{bash}
ffkmer.py -h
```

### Ex-1

Calculate all 100-mers.

```{bash}
gunzip -c data/oantigen.[1-2].gbk.gz 2> /dev/null | ffkmer.py - 100 -fmt gb | head | grep -A 2 \>
```

